packages/stream/index.d.ts

```diff
@@ -23,6 +23,34 @@ export function makePipe<
   Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
 ];
 
```

Comment:
All the problems I'm about to mention are problem I would have expected `yarn lint` to have complained about. But I pulled the `kris-stream-pump` fork, and `yarn lint` succeeds silently. This is ALARMING! What alerted me to investigate was that vscode gave me a bunch of red underlines. These differences remain even after I tell vscode to use the same version of TS that endo uses.

This file imports './types.js'. But there is no './types.js'. What is clearly intended is 'types.d.ts'

After making that fix, I get a complaint that the `import` comes before the export. Quite right.

Are the function declarations in this file the correct TS way to declare a function type rather to define a function? Clearly it must be because there are no bodies. However, some of the complaints I get (again, vscode only) would only make sense for function definitions.

For parameter names, it complains that they aren't used in the (non-existent) body.

I know that TS supports function overload. However, for `prime` it instead complains about a duplicate definition.

Reply:
I can fix the import/export order. Apart from that, cc @mhofman for clues. I’m not sure I can account for differences between the language server protocol and what `tsc` reports, but only `tsc` and `eslint` are shared objective reality with CI.

Reply:
I can say that `ses-types-test` is considerably more useful for validating our type definitions than `tsc` on a package’s internals!

Reply:
FWIW, the following diff comes from taking each of the vscode popup suggestions. It made all the red go away, except for the overload warning on `prime`. While making these changes, I ran `yarn lint` every time to see if it was ok. It was *always* ok, which is still alarming. But obviously, this PR need not wait on any of this. I am *not* suggesting you adopt the changes below. Though feel free to if any look like a genuine improvement.

```js
diff --git a/packages/stream/index.d.ts b/packages/stream/index.d.ts
index 37a19483..3e007f64 100644
--- a/packages/stream/index.d.ts
+++ b/packages/stream/index.d.ts
@@ -1,5 +1,6 @@
-export * from './types.js';
-import { AsyncQueue, Stream, Reader, Writer } from './types.js';
+import { AsyncQueue, Stream } from './types.d.ts';
+
+export * from './types.d.ts';
 
 export function makeQueue<TValue>(): AsyncQueue<TValue>;
 
@@ -7,17 +8,17 @@ export function makeStream<
   TRead,
   TWrite = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 >(
-  acks: AsyncQueue<IteratorResult<TRead, TReadReturn>>,
-  data: AsyncQueue<IteratorResult<TWrite, TWriteReturn>>,
+  _acks: AsyncQueue<IteratorResult<TRead, TReadReturn>>,
+  _data: AsyncQueue<IteratorResult<TWrite, TWriteReturn>>,
 ): Stream<TRead, TWrite, TReadReturn, TWriteReturn>;
 
 export function makePipe<
   TRead,
   TWrite = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 >(): [
   Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
   Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
@@ -27,28 +28,22 @@ export function pump<
   TRead,
   TWrite = unknown,
   TReadReturn = unknown,
-  TWriteReturn = unknown,
+  TWriteReturn = unknown
 >(
-  writer: Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
-  reader: Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
-  primer?: TWrite,
+  _writer: Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
+  _reader: Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
+  _primer?: TWrite,
 ): Promise<void>;
 
:...skipping...
diff --git a/packages/stream/index.d.ts b/packages/stream/index.d.ts
index 37a19483..3e007f64 100644
--- a/packages/stream/index.d.ts
+++ b/packages/stream/index.d.ts
@@ -1,5 +1,6 @@
-export * from './types.js';
-import { AsyncQueue, Stream, Reader, Writer } from './types.js';
+import { AsyncQueue, Stream } from './types.d.ts';
+
+export * from './types.d.ts';
 
 export function makeQueue<TValue>(): AsyncQueue<TValue>;
 
@@ -7,17 +8,17 @@ export function makeStream<
   TRead,
   TWrite = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 >(
-  acks: AsyncQueue<IteratorResult<TRead, TReadReturn>>,
-  data: AsyncQueue<IteratorResult<TWrite, TWriteReturn>>,
+  _acks: AsyncQueue<IteratorResult<TRead, TReadReturn>>,
+  _data: AsyncQueue<IteratorResult<TWrite, TWriteReturn>>,
 ): Stream<TRead, TWrite, TReadReturn, TWriteReturn>;
 
 export function makePipe<
   TRead,
   TWrite = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 >(): [
   Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
   Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
@@ -27,28 +28,22 @@ export function pump<
   TRead,
   TWrite = unknown,
   TReadReturn = unknown,
-  TWriteReturn = unknown,
+  TWriteReturn = unknown
 >(
-  writer: Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
-  reader: Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
-  primer?: TWrite,
+  _writer: Stream<TWrite, TRead, TWriteReturn, TReadReturn>,
+  _reader: Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
+  _primer?: TWrite,
 ): Promise<void>;
 
-export function prime<
-  TRead,
-  TWrite = unknown,
-  TReturn = unknown,
->(
-  writer: AsyncGenerator<TRead,  TReturn, TWrite>,
-  primer: TWrite,
+export function prime<TRead, TWrite = unknown, TReturn = unknown>(
+  _writer: AsyncGenerator<TRead, TReturn, TWrite>,
+  _primer: TWrite,
 ): Stream<TRead, TWrite, TReturn, TReturn>;
 
 // Overload:
-export function prime<
-  TRead,
->(
-  writer: AsyncGenerator<TRead, undefined, undefined>,
+export function prime<TRead>(
   // primer is implicitly undefined for this overload.
+  _writer: AsyncGenerator<TRead, undefined, undefined>,
 ): Stream<TRead, undefined, undefined, undefined>;
 
 export function mapReader<
@@ -56,10 +51,10 @@ export function mapReader<
   TReadOut = TReadIn,
   TWrite = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 >(
-  reader: Stream<TReadIn, TWrite, TReadReturn, TWriteReturn>,
-  transform: (value: TReadIn) => TReadOut,
+  _reader: Stream<TReadIn, TWrite, TReadReturn, TWriteReturn>,
+  _transform: (_value: TReadIn) => TReadOut,
 ): Stream<TReadOut, TWrite, TReadReturn, TWriteReturn>;
 
 export function mapWriter<
@@ -67,8 +62,8 @@ export function mapWriter<
   TWriteOut = TWriteIn,
   TRead = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 >(
-  writer: Stream<TRead, TWriteOut, TReadReturn, TWriteReturn>,
-  transform: (value: TWriteIn) => TWriteOut,
+  _writer: Stream<TRead, TWriteOut, TReadReturn, TWriteReturn>,
+  _transform: (_value: TWriteIn) => TWriteOut,
 ): Stream<TRead, TWriteIn, TReadReturn, TWriteReturn>;
diff --git a/packages/stream/types.d.ts b/packages/stream/types.d.ts
index cbd2d0c0..4f2d4342 100644
--- a/packages/stream/types.d.ts
+++ b/packages/stream/types.d.ts
@@ -1,6 +1,6 @@
 export interface AsyncQueue<TValue> {
-  put(value: TValue | Promise<TValue>): void,
-  get(): Promise<TValue>,
+  put(_value: TValue | Promise<TValue>): void;
+  get(): Promise<TValue>;
 }
 
 // Stream is nearly identical to AsyncGenerator and AsyncGenerator should
@@ -11,13 +11,23 @@ export interface Stream<
   TRead,
   TWrite = undefined,
   TReadReturn = undefined,
-  TWriteReturn = undefined,
+  TWriteReturn = undefined
 > {
-  next(value: TWrite): Promise<IteratorResult<TRead, TReadReturn>>,
-  return(value: TWriteReturn): Promise<IteratorResult<TRead, TReadReturn>>,
-  throw(error: Error): Promise<IteratorResult<TRead, TReadReturn>>,
-  [Symbol.asyncIterator](): Stream<TRead, TWrite, TReadReturn, TWriteReturn>,
+  next(_value: TWrite): Promise<IteratorResult<TRead, TReadReturn>>;
+  return(_value: TWriteReturn): Promise<IteratorResult<TRead, TReadReturn>>;
+  throw(_error: Error): Promise<IteratorResult<TRead, TReadReturn>>;
+  [Symbol.asyncIterator](): Stream<TRead, TWrite, TReadReturn, TWriteReturn>;
 }
 
-export type Reader<TRead, TReadReturn = undefined> = Stream<TRead, undefined, TReadReturn, undefined>;
-export type Writer<TWrite, TWriteReturn = undefined> = Stream<undefined, TWrite, undefined, TWriteReturn>;
+export type Reader<TRead, TReadReturn = undefined> = Stream<
+  TRead,
+  undefined,
+  TReadReturn,
+  undefined
+>;
+export type Writer<TWrite, TWriteReturn = undefined> = Stream<
+  undefined,
+  TWrite,
+  undefined,
+  TWriteReturn
+>;
```

Reply:
Notice both the removed commas, and the commas that were replaced by semicolons.

Reply:
> This file imports './types.js'. But there is no './types.js'. What is clearly intended is 'types.d.ts'

@erights, no, this is a quirk of TypeScript, which assumes a `.ts` ends up compiled as `.js` and `.d.ts`, and it doesn't want to rewrite imports, so it expects you still import `.js` (when using extensions). That means a `.d.ts` file shadows `.js` for anything type related (and ignore the content of the `.js` file), even when the import is `.js` everwhere.

> But I pulled the `kris-stream-pump` fork, and `yarn lint` succeeds silently. This is ALARMING!

@kriskowal, same here. After digging in, eslint by default doesn't recurse in subdirectories so `eslint .` will not check `test/`. Also by default `.ts` files are not included. That leaves only `index.js` being checked by eslint. I checked my `package.json`, and use `eslint '**/*.{js,ts}'` instead.

That said, some of the eslint complaints I see in VS code in the `.d.ts` are non-sense. Not sure why eslint is busted regarding imports in this package (I have no such issue in my runner).

> Are the function declarations in this file the correct TS way to declare a function type rather to define a function?

Actually I missed this, but `.d.ts` file should all have the `declare` keyword in front of `function` or other JavaScript declaration (not needed for `interface` or `type`).
Technically to disambiguate `import`/`export`, you should use `import type`, but that runs the risk of being incompatible with fairly old TypeScript versions (I believe those are old enough by now to ignore)

> some of the complaints I get (again, vscode only) would only make sense for function definitions.
> 
> For parameter names, it complains that they aren't used in the (non-existent) body.

Checks notes, and see that most of my `.d.ts` files have `/* eslint-disable no-unused-vars,no-redeclare */`

I believe we may need a more holistic approach, but for now it should help.

Reply:
Using `'**/*.{js,ts}'` in scripts runs us up against the problem that you can't use double quotes on unix and you can't use single quotes on Windows. It can be solved, but the cure is worse than the disease. But, `overrides: [{files: [...` seems to work and should get floated up to `@endo/eslint-config`. I’ve added it locally for now.